code/parse/sexp.cpp

```diff
@@ -2681,159 +2701,120 @@ int check_sexp_syntax(int node, int return_type, int recursive, int *bad_node, i
+							*bad_node = z;
+
+						return SEXP_CHECK_INVALID_MISSION_NAME;
+					}
+
 					// read the goal/event list from the mission file if both num_goals and num_events
 					// are < 0
-					if ((Campaign.missions[i].num_goals <= 0) && (Campaign.missions[i].num_events <= 0) )
+					if ((Campaign.missions[i].num_goals <= 0) && (Campaign.missions[i].num_events <= 0))
 						read_mission_goal_list(i);
-					
-					if (i < Campaign.num_missions) {
-						for (t=0; t<Campaign.missions[i].num_goals; t++)
+
+					if (type == OPF_GOAL_NAME) {
+						for (t = 0; t < Campaign.missions[i].num_goals; t++)
 							if (!stricmp(CTEXT(node), Campaign.missions[i].goals[t].name))
 								break;
 
 						if (t == Campaign.missions[i].num_goals)
 							return SEXP_CHECK_INVALID_GOAL_NAME;
+					} else if (type == OPF_EVENT_NAME) {
+						for (t = 0; t < Campaign.missions[i].num_events; t++)
+							if (!stricmp(CTEXT(node), Campaign.missions[i].events[t].name))
+								break;
+
+						if (t == Campaign.missions[i].num_events)
+							return SEXP_CHECK_INVALID_EVENT_NAME;
 					}
-				} else {
+				} else if (type == OPF_GOAL_NAME) {
 					// MWA -- short circuit evaluation of these things for now.
-					if ( (Operators[op].value == OP_PREVIOUS_GOAL_TRUE) || (Operators[op].value == OP_PREVIOUS_GOAL_FALSE) || (Operators[op].value == OP_PREVIOUS_GOAL_INCOMPLETE) )
+					if ((Operators[op].value == OP_PREVIOUS_GOAL_TRUE) || (Operators[op].value == OP_PREVIOUS_GOAL_FALSE) || (Operators[op].value == OP_PREVIOUS_GOAL_INCOMPLETE))
 						break;
 
-					for (i=0; i<Num_goals; i++)
+					for (i = 0; i < Num_goals; i++)
 						if (!stricmp(CTEXT(node), Mission_goals[i].name))
 							break;
 
 					if (i == Num_goals)
 						return SEXP_CHECK_INVALID_GOAL_NAME;
-				}
-
-				break;
-
-			case OPF_EVENT_NAME:
-				if ( type2 != SEXP_ATOM_STRING )
-					return SEXP_CHECK_TYPE_MISMATCH;
-
-				// like above checking for goals, check events in the campaign only if in Fred
-				// and only if in campaign mode.  Otherwise, check the current set of events
-				if ( Fred_running && (mode == SEXP_MODE_CAMPAIGN) ) {
-					z = find_parent_operator(node);
-					Assert(z >= 0);
-					z = Sexp_nodes[z].rest;  // first argument of operator should be mission name
-					Assert(z >= 0);
-					for (i=0; i<Campaign.num_missions; i++)
-						if (!stricmp(CTEXT(z), Campaign.missions[i].name))
-							break;
-
-					// read the goal/event list from the mission file if both num_goals and num_events
-					// are < 0
-					if ((Campaign.missions[i].num_goals <= 0) && (Campaign.missions[i].num_events <= 0) )
-						read_mission_goal_list(i);
-					
-					if (i < Campaign.num_missions) {
-						for (t=0; t<Campaign.missions[i].num_events; t++)
-							if (!stricmp(CTEXT(node), Campaign.missions[i].events[t].name))
-								break;
-
-						if (t == Campaign.missions[i].num_events)
-							return SEXP_CHECK_INVALID_EVENT_NAME;
-					}
-				} else {
+				} else if (type == OPF_EVENT_NAME) {
```

Comment:
If I'm reading the code right, then we'll get here if either FRED isn't running *or* we're not in campaign mode. Is that the behavior we want?

And if we're not in campaign mode, that implies single-mission mode, correct?

Reply:
Also, how about an `else` with `UNREACHABLE()` at the end of this `else-if` block, around line 2774? What would it mean for the condition at this line to be `false`?

Reply:
That's the behavior we want, because if we're not running FRED, or we are but we're not in campaign mode, then we only check the goals/events of the current mission.  And yes, non-campaign mode implies single-mission mode.

Added `UNREACHABLE` with an explanation